libraries/apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/Adapters.kt

```diff
@@ -379,8 +375,101 @@ class ObjectAdapter<T>(
       writer.writeAny(mapWriter.root()!!)
     } else {
       writer.beginObject()
-      wrappedAdapter.toJson(writer, customScalarAdapters, value,)
+      wrappedAdapter.toJson(writer, customScalarAdapters, value)
       writer.endObject()
     }
   }
-}
\ No newline at end of file
+}
+
+class CatchToResultAdapter<T>(private val wrappedAdapter: Adapter<T>) : Adapter<FieldResult<T>> {
+  override fun fromJson(reader: JsonReader, customScalarAdapters: CustomScalarAdapters): FieldResult<T> {
+    if (reader.peek() == JsonReader.Token.NULL) {
+      val error = customScalarAdapters.firstErrorStartingWith(reader.getPath())
+      if (error != null) {
+        reader.skipValue()
+        return FieldResult.Error(error)
+      }
+    }
+
+    return try {
+      FieldResult.Success(wrappedAdapter.fromJson(reader, customScalarAdapters))
+    } catch (e: ApolloGraphQLException) {
+      FieldResult.Error(e.error)
+    }
+  }
+
+  override fun toJson(writer: JsonWriter, customScalarAdapters: CustomScalarAdapters, value: FieldResult<T>) {
+    when (value) {
+      is FieldResult.Success -> wrappedAdapter.toJson(writer, customScalarAdapters, value.valueOrThrow())
+      else -> Unit // ignore errors
+    }
+  }
+}
+
+class CatchToThrowAdapter<T>(private val wrappedAdapter: Adapter<T>) : Adapter<T> {
+  override fun fromJson(reader: JsonReader, customScalarAdapters: CustomScalarAdapters): T {
+    if (reader.peek() == JsonReader.Token.NULL) {
+      val error = customScalarAdapters.firstErrorStartingWith(reader.getPath())
+      if (error != null) {
+        reader.skipValue()
+        throw ApolloGraphQLException(error)
+      }
+    }
+
+    // This may throw
+    return wrappedAdapter.fromJson(reader, customScalarAdapters)
+  }
+
+  override fun toJson(writer: JsonWriter, customScalarAdapters: CustomScalarAdapters, value: T) {
+    wrappedAdapter.toJson(writer, customScalarAdapters, value)
+  }
+}
+
+class CatchToNullAdapter<T>(private val wrappedAdapter: Adapter<T>) : Adapter<@JvmSuppressWildcards T?> {
+  override fun fromJson(reader: JsonReader, customScalarAdapters: CustomScalarAdapters): T? {
+    if (reader.peek() == JsonReader.Token.NULL) {
+      /**
+       * Code commented for reference purposes.
+       * We could check the error, but it wouldn't change the (null) result.
+       */
+//      val error = customScalarAdapters.firstErrorStartingWith(reader.getPath())
+//      if (error != null) {
+//        return null
+//      }
+      reader.skipValue()
```

Comment:
> If there's no error, shouldn't we dispatch the value to the wrapped adapter?

The value is null in that case. The `CatchToNullAdapter` replaces the `NullableAdapter` to avoid have double nullable types: `val adapter: Adapter<String??> = StringAdapter.nullable().catchToNull()` becomes just `val adapter: Adapter<String?> = StringAdapter.catchToNull()`